ledger/streaming/client/src/block_validator.rs

```diff
@@ -0,0 +1,261 @@
+            }
+
+            log::info!(
+                logger,
+                "validated {} blocks without ledger validation",
+                index
+            );
+            assert_eq!(index, 19);
+        });
+    }
+
+    #[test_with_logger]
+    fn validation_from_zero_block(logger: Logger) {
+        let simple_stream = stream::get_stream_with_n_components(20);
+        let block_validator: BlockValidator<SimpleMockStream, MockLedger> =
+            BlockValidator::new(simple_stream, None, logger.clone());
+
+        futures::executor::block_on(async move {
+            let mut stream = block_validator.get_block_stream(0).unwrap();
+            let mut index: BlockIndex = 0;
+            while let Some(data) = stream.next().await {
+                index = data.unwrap().block_data.block().index;
+            }
+
+            log::info!(
+                logger,
+                "validated {} blocks without ledger validation",
+                index
+            );
+            assert_eq!(index, 19);
+        });
+    }
+
+    #[test_with_logger]
+    fn validation_with_ledger(logger: Logger) {
+        let mut mock_ledger = get_mock_ledger(0);
+        let components = make_components(20);
+        for i in 0..2 {
+            let block_data = &components[i].block_data;
+            mock_ledger
+                .append_block(block_data.block(), block_data.contents(), None)
+                .unwrap();
+        }
+
+        let stream = SimpleMockStream::new_from_components(components);
+        let block_validator = BlockValidator::new(stream, Some(mock_ledger), logger.clone());
+
+        futures::executor::block_on(async move {
+            let mut stream = block_validator.get_block_stream(2).unwrap();
+            let mut index: BlockIndex = 0;
+            while let Some(data) = stream.next().await {
+                index = data.unwrap().block_data.block().index;
+            }
+
+            log::info!(logger, "validated {} blocks with ledger validation", index);
+            assert_eq!(index, 19);
+        });
+    }
+
+    #[test_with_logger]
+    fn invalid_previous_blocks_fail(logger: Logger) {
+        let mock_ledger = get_mock_ledger(4);
+        let quorum_set = make_quorum_set();
+        let mut components = Vec::new();
+        for i in 0..4 {
+            components.push(BlockStreamComponents::new(
+                mock_ledger.get_block_data(i).unwrap(),
+                quorum_set.clone(),
+                None,
+            ))
+        }
+
+        let stream = mock_stream_from_components(components);
+        let block_validator = BlockValidator::new(stream, Some(mock_ledger), logger.clone());
+
+        futures::executor::block_on(async move {
+            let mut stream = block_validator.get_block_stream(2).unwrap();
+            if let Some(data) = stream.next().await {
+                assert!(matches!(data, Err(StreamError::BlockValidation(_))));
```

Comment:
Optional/Equivalent: `assert_matches` prints a better error message
```suggestion
                assert_matches!(data, Err(StreamError::BlockValidation(_))));
```
(requires `use std::assert_matches`)

Reply:
I chose not to because I didn't want to include unstable features